In [3]:
!pip install ipython-sql


  Obtaining dependency information for prettytable from https://files.pythonhosted.org/packages/4d/81/316b6a55a0d1f327d04cc7b0ba9d04058cb62de6c3a4d4b0df280cbe3b0b/prettytable-3.9.0-py3-none-any.whl.metadata
  Obtaining dependency information for sqlalchemy>=2.0 from https://files.pythonhosted.org/packages/22/80/43ddb1ddeafdcbc3073c0e7a7d45b17678eeac4830c7e91bd6556527f311/SQLAlchemy-2.0.25-cp311-cp311-win_amd64.whl.metadata
  Using cached sqlparse-0.4.4-py3-none-any.whl (41 kB)
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
  Obtaining dependency information for typing-extensions>=4.6.0 from https://files.pythonhosted.org/packages/b7/f4/6a90020cd2d93349b442bfcb657d0dc91eee65491600b2cb1d388bc98e6b/typing_extensions-4.9.0-py3-none-any.whl.metadata
  Obtaining dependency information for greenlet!=0.4.17 from https://files.pythonhosted.org/packages/47/79/26d54d7d700ef65b689fc2665a40846d13e834da0486674a8d4f0f371a47/greenlet-3.0.3-cp311-cp311-win_amd64.whl.metadata
   ----


[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
import sqlite3
%load_ext sql

In [2]:
conn = sqlite3.connect('database.db')  # Replace with your database path
cursor = conn.cursor()

cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cursor.fetchall()

print(*tables, sep='\n')  # Print each table name on a new line

conn.close()

('IDS_ALLCountries_Data',)


In [3]:
%sql sqlite:///database.db

In [4]:
%sql select * from IDS_ALLCountries_Data limit 10

 * sqlite:///database.db
Done.


[('Afghanistan', 'AFG', 'World', 'WLD', 'Average grace period on new external debt commitments (years)', 'DT.GPA.DPPG', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 10.25, 8.0833, 9.9575, 0.0, 7.8333, 0.0, 0.0, 0.0, 29.8333, 0.0, 0.0, 19.8356, 0.0, 17.9041, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average grace period on new external debt commitments, official (years)', 'DT.GPA.OFFT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 10.25, 8.0833, 9.9575, 0.0, 7.8333, 0.0, 0.0, 0.0, 29.8333, 0.0, 0.0, 19.8356, 0.0, 17.9041, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average grace period on new external debt commitments, private (years)', 'DT.GPA.PRVT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average grant element on new external debt commitments (%)', 'DT.GRE.DPPG', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 50.6176, 43.0248, 58.0061, 0.0, 21.2119, 0.0, 0.0, 0.0, 81.9289, 0.0, 0.0, 64.2646, 0.0, 73.079, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average grant element on new external debt commitments, official (%)', 'DT.GRE.OFFT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 50.6176, 43.0248, 58.0061, 0.0, 21.2119, 0.0, 0.0, 0.0, 81.9289, 0.0, 0.0, 64.2646, 0.0, 73.079, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average grant element on new external debt commitments, private (%)', 'DT.GRE.PRVT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average interest on new external debt commitments (%)', 'DT.INR.DPPG', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1.2496, 1.3852, 0.4823, 0.0, 2.9999, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5705, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average interest on new external debt commitments, official (%)', 'DT.INR.OFFT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1.2496, 1.3852, 0.4823, 0.0, 2.9999, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5705, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, None, None),
 ('Afghanistan', 'AFG', 'World', 'WLD', 'Average interest on new external debt commitme

## Finding the number of distinct countries
<p>From the first ten rows, we can see the amount of debt owed by <em>Afghanistan</em> in the different debt indicators. But we do not know the number of different countries we have on the table. There are repetitions in the country names because a country is most likely to have debt in more than one debt indicator. </p>
<p>Without a count of unique countries, we will not be able to perform our statistical analyses holistically. In this section, we are going to extract the number of unique countries present in the table. </p>

In [5]:
%%sql
SELECT count(distinct "country name") AS total_distinct_countries
FROM IDS_ALLCountries_Data;

 * sqlite:///database.db
Done.


[(138,)]

## Finding out the distinct debt indicators
<p>We can see there are a total of 138 countries present on the table. As we saw in the first section, there is a column called <code>indicator_name</code> that briefly specifies the purpose of taking the debt. Just beside that column, there is another column called <code>indicator_code</code> which symbolizes the category of these debts. Knowing about these various debt indicators will help us to understand the areas in which a country can possibly be indebted to. </p>

In [15]:
%%sql
select distinct "series code" as distinct_debt_indicators from IDS_ALLCountries_Data
order by distinct_debt_indicators;

 * sqlite:///database.db
Done.


[(None,),
 ('BM.GSR.TOTL.CD',),
 ('BN.CAB.XOKA.CD',),
 ('BX.GRT.EXTA.CD.DT',),
 ('BX.GRT.TECH.CD.DT',),
 ('BX.GSR.TOTL.CD',),
 ('BX.KLT.DINV.CD.DT',),
 ('BX.KLT.DREM.CD.DT',),
 ('BX.PEF.TOTL.CD.DT',),
 ('BX.TRF.PWKR.CD.DT',),
 ('DT.AMT.BLAT.CB.CD',),
 ('DT.AMT.BLAT.CD',),
 ('DT.AMT.BLAT.GG.CD',),
 ('DT.AMT.BLAT.OPS.CD',),
 ('DT.AMT.BLAT.PRVG.CD',),
 ('DT.AMT.BLAT.PS.CD',),
 ('DT.AMT.BLTC.CB.CD',),
 ('DT.AMT.BLTC.CD',),
 ('DT.AMT.BLTC.GG.CD',),
 ('DT.AMT.BLTC.OPS.CD',),
 ('DT.AMT.BLTC.PRVG.CD',),
 ('DT.AMT.BLTC.PS.CD',),
 ('DT.AMT.DECB.CD',),
 ('DT.AMT.DEGG.CD',),
 ('DT.AMT.DEPS.CD',),
 ('DT.AMT.DIMF.CD',),
 ('DT.AMT.DLTF.CD',),
 ('DT.AMT.DLXF.CD',),
 ('DT.AMT.DOPS.CD',),
 ('DT.AMT.DPNG.CD',),
 ('DT.AMT.DPPG.CD',),
 ('DT.AMT.MIBR.CD',),
 ('DT.AMT.MIDA.CD',),
 ('DT.AMT.MLAT.CB.CD',),
 ('DT.AMT.MLAT.CD',),
 ('DT.AMT.MLAT.GG.CD',),
 ('DT.AMT.MLAT.OPS.CD',),
 ('DT.AMT.MLAT.PRVG.CD',),
 ('DT.AMT.MLAT.PS.CD',),
 ('DT.AMT.MLTC.CB.CD',),
 ('DT.AMT.MLTC.CD',),
 ('DT.AMT.MLTC.GG.CD',),
 ('DT.AMT.MLTC.OPS.CD',),
 ('DT.AMT.MLTC.PRVG.CD',),
 ('DT.AMT.MLTC.PS.CD',),
 ('DT.AMT.OFFT.CB.CD',),
 ('DT.AMT.OFFT.CD',),
 ('DT.AMT.OFFT.GG.CD',),
 ('DT.AMT.OFFT.OPS.CD',),
 ('DT.AMT.OFFT.PRVG.CD',),
 ('DT.AMT.OFFT.PS.CD',),
 ('DT.AMT.PBND.CB.CD',),
 ('DT.AMT.PBND.CD',),
 ('DT.AMT.PBND.GG.CD',),
 ('DT.AMT.PBND.OPS.CD',),
 ('DT.AMT.PBND.PRVG.CD',),
 ('DT.AMT.PBND.PS.CD',),
 ('DT.AMT.PCBK.CB.CD',),
 ('DT.AMT.PCBK.CD',),
 ('DT.AMT.PCBK.GG.CD',),
 ('DT.AMT.PCBK.OPS.CD',),
 ('DT.AMT.PCBK.PRVG.CD',),
 ('DT.AMT.PCBK.PS.CD',),
 ('DT.AMT.PNGB.CD',),
 ('DT.AMT.PNGC.CD',),
 ('DT.AMT.PROP.CB.CD',),
 ('DT.AMT.PROP.CD',),
 ('DT.AMT.PROP.GG.CD',),
 ('DT.AMT.PROP.OPS.CD',),
 ('DT.AMT.PROP.PRVG.CD',),
 ('DT.AMT.PROP.PS.CD',),
 ('DT.AMT.PRPG.CD',),
 ('DT.AMT.PRVT.CB.CD',),
 ('DT.AMT.PRVT.CD',),
 ('DT.AMT.PRVT.GG.CD',),
 ('DT.AMT.PRVT.OPS.CD',),
 ('DT.AMT.PRVT.PRVG.CD',),
 ('DT.AMT.PRVT.PS.CD',),
 ('DT.AXA.DLXF.CD',),
 ('DT.AXA.OFFT.CD',),
 ('DT.AXA.PRVT.CD',),
 ('DT.AXF.DPPG.CD',),
 ('DT.AXR.DPPG.CD',),
 ('DT.AXR.OFFT.CD',),
 ('DT.AXR.PRVT.CD',),
 ('DT.COM.BLAT.CD',),
 ('DT.COM.DPPG.CD',),
 ('DT.COM.MIBR.CD',),
 ('DT.COM.MIDA.CD',),
 ('DT.COM.MLAT.CD',),
 ('DT.COM.OFFT.CD',),
 ('DT.COM.PRVT.CD',),
 ('DT.CUR.DMAK.ZS',),
 ('DT.CUR.EURO.ZS',),
 ('DT.CUR.FFRC.ZS',),
 ('DT.CUR.JYEN.ZS',),
 ('DT.CUR.MULC.ZS',),
 ('DT.CUR.OTHC.ZS',),
 ('DT.CUR.SDRW.ZS',),
 ('DT.CUR.SWFR.ZS',),
 ('DT.CUR.UKPS.ZS',),
 ('DT.CUR.USDL.ZS',),
 ('DT.DFR.DPPG.CD',),
 ('DT.DIS.BLAT.CB.CD',),
 ('DT.DIS.BLAT.CD',),
 ('DT.DIS.BLAT.GG.CD',),
 ('DT.DIS.BLAT.OPS.CD',),
 ('DT.DIS.BLAT.PRVG.CD',),
 ('DT.DIS.BLAT.PS.CD',),
 ('DT.DIS.BLTC.CB.CD',),
 ('DT.DIS.BLTC.CD',),
 ('DT.DIS.BLTC.GG.CD',),
 ('DT.DIS.BLTC.OPS.CD',),
 ('DT.DIS.BLTC.PRVG.CD',),
 ('DT.DIS.BLTC.PS.CD',),
 ('DT.DIS.DECB.CD',),
 ('DT.DIS.DEGG.CD',),
 ('DT.DIS.DEPS.CD',),
 ('DT.DIS.DIMF.CD',),
 ('DT.DIS.DLTF.CD',),
 ('DT.DIS.DLXF.CD',),
 ('DT.DIS.DOPS.CD',),
 ('DT.DIS.DPNG.CD',),
 ('DT.DIS.DPPG.CD',),
 ('DT.DIS.IDAG.CD',),
 ('DT.DIS.MIBR.CD',),
 ('DT.DIS.MIDA.CD',),
 ('DT.DIS.MLAT.CB.CD',),
 ('DT.DIS.MLAT.CD',),
 ('DT.DIS.MLAT.GG.CD',),
 ('DT.DIS.MLAT.OPS.CD',),
 ('DT.DIS.MLAT.PRVG.CD',),
 ('DT.DIS.MLAT.PS.CD',),
 ('DT.DIS.MLTC.CB.CD',),
 ('DT.DIS.MLTC.CD',),
 ('DT.DIS.MLTC.GG.CD',),
 ('DT.DIS.MLTC.OPS.CD',),
 ('DT.DIS.MLTC.PRVG.CD',),
 ('DT.DIS.MLTC.PS.CD',),
 ('DT.DIS.OFFT.CB.CD',),
 ('DT.DIS.OFFT.CD',),
 ('DT.DIS.OFFT.GG.CD',),
 ('DT.DIS.OFFT.OPS.CD',),
 ('DT.DIS.OFFT.PRVG.CD',),
 ('DT.DIS.OFFT.PS.CD',),
 ('DT.DIS.PBND.CB.CD',),
 ('DT.DIS.PBND.CD',),
 ('DT.DIS.PBND.GG.CD',),
 ('DT.DIS.PBND.OPS.CD',),
 ('DT.DIS.PBND.PRVG.CD',),
 ('DT.DIS.PBND.PS.CD',),
 ('DT.DIS.PCBK.CB.CD',),
 ('DT.DIS.PCBK.CD',),
 ('DT.DIS.PCBK.GG.CD',),
 ('DT.DIS.PCBK.OPS.CD',),
 ('DT.DIS.PCBK.PRVG.CD',),
 ('DT.DIS.PCBK.PS.CD',),
 ('DT.DIS.PNGB.CD',),
 ('DT.DIS.PNGC.CD',),
 ('DT.DIS.PROP.CB.CD',),
 ('DT.DIS.PROP.CD',),
 ('DT.DIS.PROP.GG.CD',),
 ('DT.DIS.PROP.OPS.CD',),
 ('DT.DIS.PROP.PRVG.CD',),
 ('DT.DIS.PROP.PS.CD',),
 ('DT.DIS.PRPG.CD',),
 ('DT.DIS.PRVT.CB.CD',),
 ('DT.DIS.P

In [17]:
%%sql
select count(distinct "series code") as count_distinct_codes from IDS_ALLCountries_Data;

 * sqlite:///database.db
Done.


[(572,)]

## Totaling the amount of debt owed by the countries
<p>As mentioned earlier, the financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how we will approach it?</p>
<p>Let's switch gears from the debt indicators now and find out the total amount of debt (in USD) that is owed by the different countries. This will give us a sense of how the overall economy of the entire world is holding up.</p>

In [43]:
%%sql
SELECT 
    "Country Name" as Country,
    ROUND(
        SUM(COALESCE(NULLIF("1970", ''), 0) +
            COALESCE(NULLIF("1971", ''), 0) +
            COALESCE(NULLIF("1972", ''), 0) +
            COALESCE(NULLIF("1973", ''), 0) +
            COALESCE(NULLIF("1974", ''), 0) +
            COALESCE(NULLIF("1975", ''), 0) +
            COALESCE(NULLIF("1976", ''), 0) +
            COALESCE(NULLIF("1977", ''), 0) +
            COALESCE(NULLIF("1978", ''), 0) +
            COALESCE(NULLIF("1979", ''), 0) +
            COALESCE(NULLIF("1980", ''), 0) +
            COALESCE(NULLIF("1981", ''), 0) +
            COALESCE(NULLIF("1982", ''), 0) +
            COALESCE(NULLIF("1983", ''), 0) +
            COALESCE(NULLIF("1984", ''), 0) +
            COALESCE(NULLIF("1985", ''), 0) +
            COALESCE(NULLIF("1986", ''), 0) +
            COALESCE(NULLIF("1987", ''), 0) +
            COALESCE(NULLIF("1988", ''), 0) +
            COALESCE(NULLIF("1989", ''), 0) +
            COALESCE(NULLIF("1990", ''), 0) +
            COALESCE(NULLIF("1991", ''), 0) +
            COALESCE(NULLIF("1992", ''), 0) +
            COALESCE(NULLIF("1993", ''), 0) +
            COALESCE(NULLIF("1994", ''), 0) +
            COALESCE(NULLIF("1995", ''), 0) +
            COALESCE(NULLIF("1996", ''), 0) +
            COALESCE(NULLIF("1997", ''), 0) +
            COALESCE(NULLIF("1998", ''), 0) +
            COALESCE(NULLIF("1999", ''), 0) +
            COALESCE(NULLIF("2000", ''), 0) +
            COALESCE(NULLIF("2001", ''), 0) +
            COALESCE(NULLIF("2002", ''), 0) +
            COALESCE(NULLIF("2003", ''), 0) +
            COALESCE(NULLIF("2004", ''), 0) +
            COALESCE(NULLIF("2005", ''), 0) +
            COALESCE(NULLIF("2006", ''), 0) +
            COALESCE(NULLIF("2007", ''), 0) +
            COALESCE(NULLIF("2008", ''), 0) +
            COALESCE(NULLIF("2009", ''), 0) +
            COALESCE(NULLIF("2010", ''), 0) +
            COALESCE(NULLIF("2011", ''), 0) +
            COALESCE(NULLIF("2012", ''), 0) +
            COALESCE(NULLIF("2013", ''), 0) +
            COALESCE(NULLIF("2014", ''), 0) +
            COALESCE(NULLIF("2015", ''), 0) +
            COALESCE(NULLIF("2016", ''), 0) +
            COALESCE(NULLIF("2017", ''), 0) +
            COALESCE(NULLIF("2018", ''), 0) +
            COALESCE(NULLIF("2019", ''), 0) +
            COALESCE(NULLIF("2020", ''), 0) +
            COALESCE(NULLIF("2021", ''), 0) +
            COALESCE(NULLIF("2022", ''), 0) +
            COALESCE(NULLIF("2023", ''), 0) +
            COALESCE(NULLIF("2024", ''), 0) +
            COALESCE(NULLIF("2025", ''), 0) +
            COALESCE(NULLIF("2026", ''), 0) +
            COALESCE(NULLIF("2027", ''), 0) +
            COALESCE(NULLIF("2028", ''), 0) +
            COALESCE(NULLIF("2029", ''), 0) +
            COALESCE(NULLIF("2030", ''), 0)
        ) / 1000000, 2) AS total_debt
FROM IDS_ALLCountries_Data
GROUP BY "Country Name";


 * sqlite:///database.db
Done.


[(None, 0.0),
 ('Afghanistan', 1095863.47),
 ('Albania', 2640183.15),
 ('Algeria', 23523169.56),
 ('Angola', 21102407.67),
 ('Argentina', 88752384.31),
 ('Armenia', 2822122.11),
 ('Azerbaijan', 5983513.09),
 ('Bangladesh', 28346695.35),
 ('Belarus', 10243905.91),
 ('Belize', 684034.97),
 ('Benin', 2240578.69),
 ('Bhutan', 828748.95),
 ('Bolivia', 6602878.83),
 ('Bosnia and Herzegovina', 3848320.39),
 ('Botswana', 1727211.02),
 ('Brazil', 219416389.03),
 ('Bulgaria', 14733053.65),
 ('Burkina Faso', 2990661.07),
 ('Burundi', 780041.59),
 ('Cabo Verde', 687096.72),
 ('Cambodia', 4418930.58),
 ('Cameroon', 6838803.35),
 ('Central African Republic', 595955.21),
 ('Chad', 1624885.53),
 ('China', 554296885.59),
 ('Colombia', 47133857.78),
 ('Comoros', 217161.65),
 ('Congo, Dem. Rep.', 7030663.97),
 ('Congo, Rep.', 4221310.86),
 ('Costa Rica', 8297542.54),
 ("Cote d'Ivoire", 11730027.34),
 ('Data from database: International Debt Statistics', 0.0),
 ('Djibouti', 755416.83),
 ('Dominica', 170483.0),
 ('Dominican Republic', 11940537.95),
 ('East Asia & Pacific (excluding high income)', 856720046.33),
 ('Ecuador', 18341758.25),
 ('Egypt, Arab Rep.', 46033971.27),
 ('El Salvador', 6281471.4),
 ('Eritrea', 385177.55),
 ('Eswatini', 610071.6),
 ('Ethiopia', 11354265.84),
 ('Europe & Central Asia (excluding high income)', 466221384.22),
 ('Fiji', 758040.71),
 ('Gabon', 3765700.69),
 ('Gambia, The', 490617.36),
 ('Georgia', 4161444.62),
 ('Ghana', 11056372.48),
 ('Grenada', 281709.97),
 ('Guatemala', 7887469.87),
 ('Guinea', 2588563.39),
 ('Guinea-Bissau', 541145.18),
 ('Guyana', 1330250.14),
 ('Haiti', 1626580.59),
 ('Honduras', 5099443.0),
 ('IDA only', 224314621.03),
 ('IDA total', 343489942.56),
 ('India', 194933955.45),
 ('Indonesia', 131641573.57),
 ('Iran, Islamic Rep.', 20359819.51),
 ('Iraq', 6513287.92),
 ('Jamaica', 6101038.5),
 ('Jordan', 9723274.28),
 ('Kazakhstan', 33662722.19),
 ('Kenya', 11935067.17),
 ('Kosovo', 569700.06),
 ('Kyrgyz Republic', 2180287.07),
 ('Lao PDR', 4252270.73),
 ('Last Updated: 12/13/2023', 0.0),
 ('Latin America & Caribbean (excluding high income)', 679763996.0),
 ('Least developed countries: UN classification', 178164348.63),
 ('Lebanon', 18030509.78),
 ('Lesotho', 803647.61),
 ('Liberia', 1069208.77),
 ('Low & middle income', 2753475872.83),
 ('Low income', 84218397.35),
 ('Lower middle income', 757424121.33),
 ('Madagascar', 3232546.13),
 ('Malawi', 1923156.95),
 ('Maldives', 945751.06),
 ('Mali', 3076039.58),
 ('Mauritania', 2385980.07),
 ('Mauritius', 3314633.62),
 ('Mexico', 211157592.66),
 ('Middle East & North Africa (excluding high income)', 184852297.55),
 ('Middle income', 2665937529.78),
 ('Moldova', 1675862.04),
 ('Mongolia', 4833304.1),
 ('Montenegro', 1611509.6),
 ('Morocco', 25460013.66),
 ('Mozambique', 8631834.67),
 ('Myanmar', 7195376.76),
 ('Nepal', 3886299.36),
 ('Nicaragua', 5793204.12),
 ('Niger', 2117574.5),
 ('Nigeria', 34274541.81),
 ('North Macedonia', 2581399.71),
 ('Pakistan', 42187114.34),
 ('Papua New Guinea', 4480973.56),
 ('Paraguay', 5854894.9),
 ('Peru', 27640868.76),
 ('Philippines', 46611326.95),
 ('Russian Federation', 187088800.96),
 ('Rwanda', 2025060.78),
 ('Samoa', 217941.49),
 ('Sao Tome and Principe', 174608.89),
 ('Senegal', 6179120.73),
 ('Serbia', 15619635.46),
 ('Sierra Leone', 1031758.67),
 ('Solomon Islands', 184695.82),
 ('Somalia', 1618720.97),
 ('South Africa', 47233277.39),
 ('South Asia', 288359098.81),
 ('Sri Lanka', 16082715.87),
 ('St. Lucia', 329687.61),
 ('St. Vincent and the Grenadines', 213594.95),
 ('Sub-Saharan Africa (excluding high income)', 261424534.62),
 ('Sudan', 10889261.92),
 ('Suriname', 512555.43),
 ('Syrian Arab Republic', 3241045.84),
 ('Tajikistan', 1574307.31),
 ('Tanzania', 9655194.29),
 ('Thailand', 59953727.24),
 ('Timor-Leste', 189074.07),
 ('Togo', 1380285.32),
 ('Tonga', 97340.82),
 ('Tunisia', 13864071.94),
 ('Turkiye', 124629150.5),
 ('Turkmenistan', 2741311.63),
 ('Uganda', 5171244.34),
 ('Ukraine', 33003720.97),
 ('U

## Country with the highest debt
<p>"Human beings cannot comprehend very large or very small numbers. It would be useful for us to acknowledge that fact." - <a href="https://en.wikipedia.org/wiki/Daniel_Kahneman">Daniel Kahneman</a>. That is more than <em>3 million <strong>million</strong></em> USD, an amount which is really hard for us to fathom. </p>
<p>Now that we have the exact total of the amounts of debt owed by several countries, let's now find out the country that owns the highest amount of debt along with the amount. <strong>Note</strong> that this debt is the sum of different debts owed by a country across several categories. This will help to understand more about the country in terms of its socio-economic scenarios. We can also find out the category in which the country owns its highest debt. But we will leave that for now. </p>

In [59]:
%%sql
WITH TotalDebtCTE AS (
    SELECT 
        "Country Name" as Country,
        ROUND(
            SUM(
                COALESCE(NULLIF("1970", ''), 0) +
                COALESCE(NULLIF("1971", ''), 0) +
                COALESCE(NULLIF("1972", ''), 0) +
                COALESCE(NULLIF("1973", ''), 0) +
                COALESCE(NULLIF("1974", ''), 0) +
                COALESCE(NULLIF("1975", ''), 0) +
                COALESCE(NULLIF("1976", ''), 0) +
                COALESCE(NULLIF("1977", ''), 0) +
                COALESCE(NULLIF("1978", ''), 0) +
                COALESCE(NULLIF("1979", ''), 0) +
                COALESCE(NULLIF("1980", ''), 0) +
                COALESCE(NULLIF("1981", ''), 0) +
                COALESCE(NULLIF("1982", ''), 0) +
                COALESCE(NULLIF("1983", ''), 0) +
                COALESCE(NULLIF("1984", ''), 0) +
                COALESCE(NULLIF("1985", ''), 0) +
                COALESCE(NULLIF("1986", ''), 0) +
                COALESCE(NULLIF("1987", ''), 0) +
                COALESCE(NULLIF("1988", ''), 0) +
                COALESCE(NULLIF("1989", ''), 0) +
                COALESCE(NULLIF("1990", ''), 0) +
                COALESCE(NULLIF("1991", ''), 0) +
                COALESCE(NULLIF("1992", ''), 0) +
                COALESCE(NULLIF("1993", ''), 0) +
                COALESCE(NULLIF("1994", ''), 0) +
                COALESCE(NULLIF("1995", ''), 0) +
                COALESCE(NULLIF("1996", ''), 0) +
                COALESCE(NULLIF("1997", ''), 0) +
                COALESCE(NULLIF("1998", ''), 0) +
                COALESCE(NULLIF("1999", ''), 0) +
                COALESCE(NULLIF("2000", ''), 0) +
                COALESCE(NULLIF("2001", ''), 0) +
                COALESCE(NULLIF("2002", ''), 0) +
                COALESCE(NULLIF("2003", ''), 0) +
                COALESCE(NULLIF("2004", ''), 0) +
                COALESCE(NULLIF("2005", ''), 0) +
                COALESCE(NULLIF("2006", ''), 0) +
                COALESCE(NULLIF("2007", ''), 0) +
                COALESCE(NULLIF("2008", ''), 0) +
                COALESCE(NULLIF("2009", ''), 0) +
                COALESCE(NULLIF("2010", ''), 0) +
                COALESCE(NULLIF("2011", ''), 0) +
                COALESCE(NULLIF("2012", ''), 0) +
                COALESCE(NULLIF("2013", ''), 0) +
                COALESCE(NULLIF("2014", ''), 0) +
                COALESCE(NULLIF("2015", ''), 0) +
                COALESCE(NULLIF("2016", ''), 0) +
                COALESCE(NULLIF("2017", ''), 0) +
                COALESCE(NULLIF("2018", ''), 0) +
                COALESCE(NULLIF("2019", ''), 0) +
                COALESCE(NULLIF("2020", ''), 0) +
                COALESCE(NULLIF("2021", ''), 0) +
                COALESCE(NULLIF("2022", ''), 0) +
                COALESCE(NULLIF("2023", ''), 0) +
                COALESCE(NULLIF("2024", ''), 0) +
                COALESCE(NULLIF("2025", ''), 0) +
                COALESCE(NULLIF("2026", ''), 0) +
                COALESCE(NULLIF("2027", ''), 0) +
                COALESCE(NULLIF("2028", ''), 0) +
                COALESCE(NULLIF("2029", ''), 0) +
                COALESCE(NULLIF("2030", ''), 0)
            ) / 1000000, 2) AS total_debt
    FROM IDS_ALLCountries_Data
    GROUP BY "Country Name"
)
SELECT 
    Country,
    total_debt
FROM TotalDebtCTE
WHERE Country IS NOT NULL  -- Exclude rows where Country is NULL
ORDER BY total_debt DESC
LIMIT 10;

 * sqlite:///database.db
Done.


[('Low & middle income', 2753475872.83),
 ('Middle income', 2665937529.78),
 ('Upper middle income', 1909881949.88),
 ('East Asia & Pacific (excluding high income)', 856720046.33),
 ('Lower middle income', 757424121.33),
 ('Latin America & Caribbean (excluding high income)', 679763996.0),
 ('China', 554296885.59),
 ('Europe & Central Asia (excluding high income)', 466221384.22),
 ('IDA total', 343489942.56),
 ('South Asia', 288359098.81)]

## Average amount of debt across indicators
<p>So, it was <em>China</em>. A more in-depth breakdown of China's debts can be found <a href="https://datatopics.worldbank.org/debt/ids/country/CHN">here</a>. </p>
<p>We now have a brief overview of the dataset and a few of its summary statistics. We already have an idea of the different debt indicators in which the countries owe their debts. We can dig even further to find out on an average how much debt a country owes? This will give us a better sense of the distribution of the amount of debt across different indicators.</p>

In [62]:
%%sql
SELECT 
    "Series Code" AS debt_indicator,
    "Series Name" AS indicator_name,
    ROUND(AVG(
        COALESCE(NULLIF("1970", ''), 0) +
        COALESCE(NULLIF("1971", ''), 0) +
        COALESCE(NULLIF("1972", ''), 0) +
        COALESCE(NULLIF("1973", ''), 0) +
        COALESCE(NULLIF("1974", ''), 0) +
        COALESCE(NULLIF("1975", ''), 0) +
        COALESCE(NULLIF("1976", ''), 0) +
        COALESCE(NULLIF("1977", ''), 0) +
        COALESCE(NULLIF("1978", ''), 0) +
        COALESCE(NULLIF("1979", ''), 0) +
        COALESCE(NULLIF("1980", ''), 0) +
        COALESCE(NULLIF("1981", ''), 0) +
        COALESCE(NULLIF("1982", ''), 0) +
        COALESCE(NULLIF("1983", ''), 0) +
        COALESCE(NULLIF("1984", ''), 0) +
        COALESCE(NULLIF("1985", ''), 0) +
        COALESCE(NULLIF("1986", ''), 0) +
        COALESCE(NULLIF("1987", ''), 0) +
        COALESCE(NULLIF("1988", ''), 0) +
        COALESCE(NULLIF("1989", ''), 0) +
        COALESCE(NULLIF("1990", ''), 0) +
        COALESCE(NULLIF("1991", ''), 0) +
        COALESCE(NULLIF("1992", ''), 0) +
        COALESCE(NULLIF("1993", ''), 0) +
        COALESCE(NULLIF("1994", ''), 0) +
        COALESCE(NULLIF("1995", ''), 0) +
        COALESCE(NULLIF("1996", ''), 0) +
        COALESCE(NULLIF("1997", ''), 0) +
        COALESCE(NULLIF("1998", ''), 0) +
        COALESCE(NULLIF("1999", ''), 0) +
        COALESCE(NULLIF("2000", ''), 0) +
        COALESCE(NULLIF("2001", ''), 0) +
        COALESCE(NULLIF("2002", ''), 0) +
        COALESCE(NULLIF("2003", ''), 0) +
        COALESCE(NULLIF("2004", ''), 0) +
        COALESCE(NULLIF("2005", ''), 0) +
        COALESCE(NULLIF("2006", ''), 0) +
        COALESCE(NULLIF("2007", ''), 0) +
        COALESCE(NULLIF("2008", ''), 0) +
        COALESCE(NULLIF("2009", ''), 0) +
        COALESCE(NULLIF("2010", ''), 0) +
        COALESCE(NULLIF("2011", ''), 0) +
        COALESCE(NULLIF("2012", ''), 0) +
        COALESCE(NULLIF("2013", ''), 0) +
        COALESCE(NULLIF("2014", ''), 0) +
        COALESCE(NULLIF("2015", ''), 0) +
        COALESCE(NULLIF("2016", ''), 0) +
        COALESCE(NULLIF("2017", ''), 0) +
        COALESCE(NULLIF("2018", ''), 0) +
        COALESCE(NULLIF("2019", ''), 0) +
        COALESCE(NULLIF("2020", ''), 0) +
        COALESCE(NULLIF("2021", ''), 0) +
        COALESCE(NULLIF("2022", ''), 0) +
        COALESCE(NULLIF("2023", ''), 0) +
        COALESCE(NULLIF("2024", ''), 0) +
        COALESCE(NULLIF("2025", ''), 0) +
        COALESCE(NULLIF("2026", ''), 0) +
        COALESCE(NULLIF("2027", ''), 0) +
        COALESCE(NULLIF("2028", ''), 0) +
        COALESCE(NULLIF("2029", ''), 0) +
        COALESCE(NULLIF("2030", ''), 0)
    ) / COUNT(DISTINCT "Country Name"), 2) AS average_debt
FROM IDS_ALLCountries_Data
GROUP BY debt_indicator, indicator_name
ORDER BY average_debt DESC
LIMIT 10;


 * sqlite:///database.db
Done.


[('NY.GNP.MKTP.CD', 'GNI (current US$)', 151480228107.27),
 ('DT.DOD.DECT.CD', 'External debt stocks, total (DOD, current US$)', 40585855445.79),
 ('BM.GSR.TOTL.CD', 'Imports of goods, services and primary income (current US$)', 39358680855.24),
 ('BX.GSR.TOTL.CD', 'Exports of goods, services and primary income (current US$)', 37014029046.62),
 ('DT.DOD.DLXF.CD', 'External debt stocks, long-term (DOD, current US$)', 30232286479.4),
 ('FI.RES.TOTL.CD', 'Total reserves (includes gold, current US$)', 29406274682.19),
 ('DT.DOD.DPPG.CD', 'External debt stocks, public and publicly guaranteed (PPG) (DOD, current US$)', 18541669492.38),
 ('DT.DOD.DEPS.CD', 'External debt stocks, public sector (PPG) (DOD, current US$)', 18467819130.12),
 ('DT.DOD.PUBS.CD', 'External debt stocks, long-term public sector (DOD, current US$)', 18463883613.47),
 ('DT.DOD.VTOT.CD', 'External debt stocks, variable rate (DOD, current US$)', 17509330142.15)]

## The highest amount of principal repayments
<p>We can see that the indicator <code>NY.GNP.MKTP.CD</code> tops the chart of average debt. This category includes repayment of long term debts. Countries take on long-term debt to acquire immediate capital. More information about this category can be found <a href="https://datacatalog.worldbank.org/principal-repayments-external-debt-long-term-amt-current-us-0">here</a>. </p>
<p>An interesting observation in the above finding is that there is a huge difference in the amounts of the indicators after the second one. This indicates that the first two indicators might be the most severe categories in which the countries owe their debts.</p>
<p>We can investigate this a bit more so as to find out which country owes the highest amount of debt in the category of long term debts (<code>NY.GNP.MKTP.CD</code>). Since not all the countries suffer from the same kind of economic disturbances, this finding will allow us to understand that particular country's economic condition a bit more specifically. </p>

In [68]:
%%sql
WITH TotalDebtCTE AS (
    SELECT 
        "Country Name" as Country,
        "Series Code",
        ROUND(
            SUM(
                COALESCE(NULLIF("1970", ''), 0) +
                COALESCE(NULLIF("1971", ''), 0) +
                COALESCE(NULLIF("1972", ''), 0) +
                COALESCE(NULLIF("1973", ''), 0) +
                COALESCE(NULLIF("1974", ''), 0) +
                COALESCE(NULLIF("1975", ''), 0) +
                COALESCE(NULLIF("1976", ''), 0) +
                COALESCE(NULLIF("1977", ''), 0) +
                COALESCE(NULLIF("1978", ''), 0) +
                COALESCE(NULLIF("1979", ''), 0) +
                COALESCE(NULLIF("1980", ''), 0) +
                COALESCE(NULLIF("1981", ''), 0) +
                COALESCE(NULLIF("1982", ''), 0) +
                COALESCE(NULLIF("1983", ''), 0) +
                COALESCE(NULLIF("1984", ''), 0) +
                COALESCE(NULLIF("1985", ''), 0) +
                COALESCE(NULLIF("1986", ''), 0) +
                COALESCE(NULLIF("1987", ''), 0) +
                COALESCE(NULLIF("1988", ''), 0) +
                COALESCE(NULLIF("1989", ''), 0) +
                COALESCE(NULLIF("1990", ''), 0) +
                COALESCE(NULLIF("1991", ''), 0) +
                COALESCE(NULLIF("1992", ''), 0) +
                COALESCE(NULLIF("1993", ''), 0) +
                COALESCE(NULLIF("1994", ''), 0) +
                COALESCE(NULLIF("1995", ''), 0) +
                COALESCE(NULLIF("1996", ''), 0) +
                COALESCE(NULLIF("1997", ''), 0) +
                COALESCE(NULLIF("1998", ''), 0) +
                COALESCE(NULLIF("1999", ''), 0) +
                COALESCE(NULLIF("2000", ''), 0) +
                COALESCE(NULLIF("2001", ''), 0) +
                COALESCE(NULLIF("2002", ''), 0) +
                COALESCE(NULLIF("2003", ''), 0) +
                COALESCE(NULLIF("2004", ''), 0) +
                COALESCE(NULLIF("2005", ''), 0) +
                COALESCE(NULLIF("2006", ''), 0) +
                COALESCE(NULLIF("2007", ''), 0) +
                COALESCE(NULLIF("2008", ''), 0) +
                COALESCE(NULLIF("2009", ''), 0) +
                COALESCE(NULLIF("2010", ''), 0) +
                COALESCE(NULLIF("2011", ''), 0) +
                COALESCE(NULLIF("2012", ''), 0) +
                COALESCE(NULLIF("2013", ''), 0) +
                COALESCE(NULLIF("2014", ''), 0) +
                COALESCE(NULLIF("2015", ''), 0) +
                COALESCE(NULLIF("2016", ''), 0) +
                COALESCE(NULLIF("2017", ''), 0) +
                COALESCE(NULLIF("2018", ''), 0) +
                COALESCE(NULLIF("2019", ''), 0) +
                COALESCE(NULLIF("2020", ''), 0) +
                COALESCE(NULLIF("2021", ''), 0) +
                COALESCE(NULLIF("2022", ''), 0) +
                COALESCE(NULLIF("2023", ''), 0) +
                COALESCE(NULLIF("2024", ''), 0) +
                COALESCE(NULLIF("2025", ''), 0) +
                COALESCE(NULLIF("2026", ''), 0) +
                COALESCE(NULLIF("2027", ''), 0) +
                COALESCE(NULLIF("2028", ''), 0) +
                COALESCE(NULLIF("2029", ''), 0) +
                COALESCE(NULLIF("2030", ''), 0)
            ) / 1000000, 2) AS total_debt
    FROM IDS_ALLCountries_Data
    GROUP BY "Country Name", "Series Code"
)
SELECT 
    Country,
    total_debt
FROM TotalDebtCTE
WHERE "Series Code" = "NY.GNP.MKTP.CD" -- Exclude rows where Country is NULL
ORDER BY total_debt DESC
LIMIT 5;


 * sqlite:///database.db
Done.


[('Low & middle income', 548332621.0),
 ('Middle income', 537448671.0),
 ('Upper middle income', 412609069.0),
 ('East Asia & Pacific (excluding high income)', 232437109.0),
 ('China', 191288827.24)]

## The most common debt indicator
<p>China has the highest amount of debt in the long-term debt (<code>NY.GNP.MKTP.CD</code>) category. This is verified by <a href="https://data.worldbank.org/indicator/DT.AMT.DLXF.CD?end=2018&most_recent_value_desc=true">The World Bank</a>. It is often a good idea to verify our analyses like this since it validates that our investigations are correct. </p>
<p>We saw that long-term debt is the topmost category when it comes to the average amount of debt. But is it the most common indicator in which the countries owe their debt? Let's find that out. </p>

In [78]:
%%sql
select "Series Code" as indicator_code, count("Series Code") as indicator_count from IDS_ALLCountries_Data
group by indicator_code
order by indicator_count, indicator_code desc
limit 10;


 * sqlite:///database.db
Done.


[(None, 0),
 ('SP.POP.TOTL', 136),
 ('NY.GNP.MKTP.CD', 136),
 ('FI.RES.TOTL.MO', 136),
 ('FI.RES.TOTL.DT.ZS', 136),
 ('FI.RES.TOTL.CD', 136),
 ('DT.UND.PRVT.CD', 136),
 ('DT.UND.OFFT.CD', 136),
 ('DT.UND.DPPG.CD', 136),
 ('DT.TXR.DPPG.CD', 136)]